xdsl/transforms/experimental/ConvertStencilToLLMLIR.py

```diff
@@ -110,32 +119,120 @@ def match_and_rewrite(self, op: LoadOp, rewriter: PatternRewriter, /):
-        entry = op.region.blocks[0]
+        body = prepare_apply_body(op, rewriter)
+        dim = len(op.lb.array.data)
 
-        for arg in entry.args:
-            arg_uses = set(arg.uses)
-            for use in arg_uses:
-                use.operation.replace_operand(use.index, op.args[use.index])
-            entry.erase_arg(arg)
+        threads_per_dim = [[1024, 32, 8][dim - 1]] * dim + [1] * (3 - dim)
+        cst_tpd = [
+            arith.Constant.from_int_and_width(tpd, builtin.IndexType())
+            for tpd in threads_per_dim
+        ]
 
-        dim = len(op.lb.array.data)
+        dims = IndexAttr.size_from_bounds(op.lb, op.ub)
+        one = arith.Constant.from_int_and_width(1, builtin.IndexType())
+
+        bounds_cst = [
+            arith.Constant.from_int_and_width(dims[i], builtin.IndexType())
+            for i in range(dim)
+        ]
+
+        blocks_divs = [
+            arith.CeilDivUI.get(bound, tpd)
+            for bound, tpd in zip(bounds_cst, cst_tpd)
+        ]
+
+        launch = gpu.LaunchOp.get(body, blocks_divs + [one] * (3 - dim),
+                                  cst_tpd)
+
+        index_compute_block = Block.from_arg_types([builtin.IndexType()] * 12)
+        block_mul = [
+            arith.Muli.get(index_compute_block.args[i],
+                           index_compute_block.args[9 + i]) for i in range(dim)
+        ]
+        thread_add = [
+            arith.Addi.get(index_compute_block.args[3 + i], block_mul[i])
+            for i in range(dim)
+        ]
+
+        thread_int = [
+            arith.IndexCastOp.get(ta, builtin.IntegerType.from_width(64))
+            for ta in thread_add
+        ]
+
+        bounds_int = [
+            arith.IndexCastOp.get(bc, builtin.IntegerType.from_width(64))
+            for bc in bounds_cst
+        ]
+
+        cmpis: list[arith.Cmpi | arith.AndI] = [
+            arith.Cmpi.from_mnemonic(ti, bc, "ult")
+            for ti, bc in zip(thread_int, bounds_int)
+        ]
+
+        if len(cmpis) >= 2:
+            cmpis.append(arith.AndI.get(cmpis[0], cmpis[1]))
+        for i in range(2, len(cmpis) - 1):
+            cmpis.append(arith.AndI.get(cmpis[i], cmpis[-1]))
+
+        else_block = Block.from_ops([gpu.TerminatorOp.get()])
+
+        body_branch = cf.ConditionalBranch.get(cmpis[-1], body.blocks[0],
+                                               list(thread_add), else_block,
+                                               [])
+
+        index_compute_block.add_ops([
+            *block_mul, *thread_add, *thread_int, *bounds_int, *cmpis,
+            body_branch
+        ])
+
+        body.insert_block(index_compute_block, 0)
+        body.add_block(else_block)
+        launch.body.blocks[1].add_op(gpu.TerminatorOp.get())
+
+        # Replace with the loop and necessary constants.
+        rewriter.insert_op_before_matched_op(
+            [*cst_tpd, *bounds_cst, *blocks_divs, one, launch])
+        rewriter.erase_matched_op(False)
```

Comment:
```suggestion
        rewriter.erase_matched_op(safe_erase=False)
```
I had to look up what `False` did, we might want to force it to be a named parameter